source/exe/main_common.cc

```diff
@@ -89,6 +104,251 @@ void MainCommonBase::adminRequest(absl::string_view path_and_query, absl::string
     handler(*response_headers, body);
   });
 }
+
+namespace {
+
+class AdminResponseImpl : public MainCommonBase::AdminResponse {
+public:
+  using CleanupFn = std::function<void(AdminResponseImpl*)>;
+
+  AdminResponseImpl(Server::Instance& server, absl::string_view path, absl::string_view method,
+                    CleanupFn cleanup)
+      : server_(server), opt_admin_(server.admin()), cleanup_(cleanup) {
+    request_headers_->setMethod(method);
+    request_headers_->setPath(path);
+  }
+
+  ~AdminResponseImpl() {
+    // MainCommonBase::response_set_ holds a raw pointer to all outstanding
+    // responses (not a shared_ptr). So when destructing the response
+    // we must call cleanup on MainCommonBase if this has not already
+    // occurred.
+    //
+    // Note it's also possible for MainCommonBase to be deleted before
+    // AdminResponseImpl, in which case it will use its response_set_
+    // to call terminate, so we'll track that here and skip calling
+    // cleanup_ in that case.
+    if (!terminated_) {
```

Comment:
Added:
```
    //
    // Note that though terminated_ is mutex-protected, the destructor
    // will not run until all other refs to this have been dropped, so
    // there can be no concurrent accesses.
```

I will push this to github when there's other changes to avoid churning CI on a comment change.

Reply:
Actually this discussion yielded some a race I needed to resolve by adding a new TemrinateNotify structure which helps manage the turn-down of the responses and server in any order.